src/PoolManager.sol

```diff
@@ -203,34 +204,50 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
         PoolId id = key.toId();
         _checkPoolInitialized(id);
 
-        key.hooks.beforeSwap(key, params, hookData);
+        int128 hookDeltaInSpecified;
+        (params.amountSpecified, hookDeltaInSpecified) = key.hooks.beforeSwap(key, params, hookData);
 
-        uint256 feeForProtocol;
-        uint24 swapFee;
-        Pool.SwapState memory state;
-        (delta, feeForProtocol, swapFee, state) = pools[id].swap(
+        // execute swap, account protocol fees, and emit swap event
+        delta = _swap(
+            id,
             Pool.SwapParams({
                 tickSpacing: key.tickSpacing,
                 zeroForOne: params.zeroForOne,
                 amountSpecified: params.amountSpecified,
                 sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
+            }),
+            params.zeroForOne ? key.currency0 : key.currency1
         );
 
-        _accountPoolBalanceDelta(key, delta);
+        (int128 hookDeltaInUnspecified) = key.hooks.afterSwap(key, params, delta, hookData);
```

Comment:
now afterSwap gets the original `params.amountSpecified` (important for custom curves to calculate the output for exactInput swaps)